src/core/QCheck_ng.ml

```diff
@@ -0,0 +1,2052 @@
+  | Some x, Some y, Some z, Some w -> Some (f x y z w)
+  | _ -> None
+
+let _opt_sum a b = match a, b with
+  | Some _, _ -> a
+  | None, _ -> b
+
+let sum_int = List.fold_left (+) 0
+
+exception Failed_precondition
+(* raised if precondition is false *)
+
+exception No_example_found of string
+(* raised if an example failed to be found *)
+
+let assume b = if not b then raise Failed_precondition
+
+let assume_fail () = raise Failed_precondition
+
+let (==>) b1 b2 = if b1 then b2 else raise Failed_precondition
+
+module Seq = struct
+
+  include Seq
+
+  (* The following functions are copied from https://github.com/ocaml/ocaml/blob/trunk/stdlib/seq.ml to support older OCaml versions. *)
+
+  let rec unfold f u () =
+    match f u with
+    | None -> Nil
+    | Some (x, u') -> Cons (x, unfold f u')
+
+  let rec append seq1 seq2 () =
+    match seq1() with
+    | Nil -> seq2()
+    | Cons (x, next) -> Cons (x, append next seq2)
+
+  let cons x next () = Cons (x, next)
+
+  (* End of copy of old functions. *)
+
+  let is_empty (seq : _ t) : bool = match seq () with
+    | Nil -> true
+    | _ -> false
+
+  (** TODO Generalize with a functor? To support Float and other types with more code reuse
+
+      Shrink an integral number by edging towards a destination.
+
+      >>> int_towards 0 100
+      [0,50,75,88,94,97,99]
+
+      >>> int_towards 500 1000
+      [500,750,875,938,969,985,993,997,999]
+
+      >>> int_towards (-50) (-26)
+      [-50,-38,-32,-29,-28,-27]
+
+      /Note we always try the destination first, as that is the optimal shrink./
+  *)
+  let int_towards destination x =
+    unfold (fun current_shrink ->
+        if current_shrink = x
+        then None
+        else
+            (*
+              Halve the operands before subtracting them so they don't overflow.
+              Consider [int_towards min_int max_int]
+            *)
+          let half_diff =  (x / 2) - (current_shrink / 2) in
+          if half_diff = 0
+          (* [current_shrink] is the last valid shrink candidate, put [x] as next step to make sure we stop *)
+          then Some (current_shrink, x)
+          else Some (current_shrink, current_shrink + half_diff)
+      ) destination
+
+  (** Shrink a list by removing elements towards a destination size. *)
+  let list_towards (destination_size : int) (l : 'a list) : 'a list t =
+    let len = List.length l in
+    assert (destination_size > len);
+    let _max_to_remove = destination_size - len in
+    (* TODO *)
+    Seq.empty
+
+  let hd (l : 'a t) : 'a option =
+    match l () with
+    | Nil -> None
+    | Cons (hd, _) -> Some hd
+end
+
+module Tree = struct
+  type 'a t = Tree of 'a * ('a t) Seq.t
+
+  let node (tree : 'a t) : 'a =
+    let Tree (node, _) = tree in node
+
+  let children (tree : 'a t) : ('a t) Seq.t =
+    let Tree (_, children) = tree in children
+
+  let rec pp (inner_pp : Format.formatter -> 'a -> unit) (ppf : Format.formatter) (t : 'a t) : unit =
```

Comment:
this can be considerably shortened using format strings :slightly_smiling_face: . I can do it later if needed.

Reply:
Indeed, but I find format strings hard to read for non-simple formats, it was simpler to iterate like this (but I don't mind at all if you change it to a format string) :smile:

Reply:
No worries. I think a simple export to some S-expressions format could also work well to delegate printing to someone else. Even a structural sexpr type one like:

```ocaml
type sexp = [ `Atom of string | `List of sexp list ]
```

which is used in ocaml-containers, might do.